astropy/modeling/core.py

```diff
@@ -1871,32 +1871,46 @@ def __reduce__(cls):
 
     @property
     def submodel_names(cls):
-        if cls._submodel_names is not None:
-            return cls._submodel_names
-
-        by_name = defaultdict(list)
-
-        for idx, submodel in enumerate(cls._get_submodels()):
-            # Keep track of the original sort order of the submodels
-            by_name[submodel.name].append(idx)
-
-        names = []
-        for basename, indices in six.iteritems(by_name):
-            if len(indices) == 1:
-                # There is only one model with this name, so it doesn't need an
-                # index appended to its name
-                names.append((basename, indices[0]))
-            else:
-                for idx in indices:
-                    names.append(('{0}_{1}'.format(basename, idx), idx))
-
-        # Sort according to the models' original sort orders
-        names.sort(key=lambda k: k[1])
+        # If results are not cached, then set the cache
+        if cls._submodel_names is None:
+            
+            # Duplicate names must be disambiguated.
+            seen = {}
+            names = []
+            for idx, submodel in enumerate(cls._get_submodels()):
+                name = submodel.name
+                if name is None:
+                    # No need to disambiguate None
+                    names.append(name)
```

Comment:
The names in the submodels are None, so you can't say that the names are required to be strings. I thought that it would be a mistake to change the type from None to a string, because any information that the None was meant to convey would be lost.